### PR TITLE
Support referrent-explicit format creating presentation request

### PR DESCRIPTION
**Pull request by Patrik-Stas**
- When creating proof request, our API expected list of `<predicate_info>` objects (as they are defined in https://hyperledger.github.io/anoncreds-spec/#create-presentation-request). Aries-vcx implementation then auto-generated referent names, eg given:
```
[
  { name: 'age', p_type: '>=', p_value: 18 }, 
  { name: 'balance', p_type: '>=', p_value: 10000 }
]
```
we transformed the input internally to anoncreds format
```
{
  predicate_0: { name: 'age', p_type: '>=', p_value: 18 }, 
  predicate_1: { name: 'balance', p_type: '>=', p_value: 10000 }
}
```
- This PR enabled API called to simply supply entire `requested_predicates` object (as per anoncreds link above) and therefore letting the caller specify referents names eg.
```
{
  is_adult: { name: 'age', p_type: '>=', p_value: 18 }, 
  credit: { name: 'balance', p_type: '>=', p_value: 10000 }
}
```
This is useful capability when inspecting presentations.

#### Note
- We should remove vector format, it's convenient but not very useful
- String parsing should be happening on layer above `aries-vcx`
- But these are for different PRs

### CODE REVIEW

**Comment by gmulhearn** [comment]
had a passing browse of this; Rust changes LGTM! Keen for the typing changes in #845 . inspired me to make this ticket https://github.com/hyperledger/aries-vcx/issues/847

File: aries_vcx/src/common/proofs/proof_request.rs [hunk 1]
```diff
@@ -36,16 +36,27 @@ impl ProofRequestData {
...
-    pub fn set_requested_attributes_as_string(mut self, requested_attrs: String) -> VcxResult<Self> {
-        match serde_json::from_str::<HashMap<String, AttrInfo>>(&requested_attrs) {
+    // Support requested_attributes in 2 formats
+    // Referent-less format, deprecated:
+    // [ { name: 'age' }, { name: 'balance' } ]
+    // Explicit referent format:
+    // { 'attr_age': { 'name': 'age' }, 'attr_balance': { 'name': 'balance' } }
+    // todo: move String parsing to libvcx_core
+    pub fn set_requested_attributes_as_string(mut self, requested_attributes: String) -> VcxResult<Self> {
+        match serde_json::from_str::<HashMap<String, AttrInfo>>(&requested_attributes) {
```
Comment (bobozaur) [inline aries_vcx/src/common/proofs/proof_request.rs:46]: I'd move the deserialization into a variable for readability.

File: aries_vcx/src/common/proofs/proof_request.rs [hunk 2]
```diff
@@ -36,16 +36,27 @@ impl ProofRequestData {
...
+                let requested_attributes: Vec<AttrInfo> =
+                    ::serde_json::from_str(&requested_attributes).map_err(|err| {
+                        AriesVcxError::from_msg(
+                            AriesVcxErrorKind::InvalidJson,
+                            format!(
+                                "Invalid Requested Attributes: {:?}, err: {:?}",
+                                requested_attributes, err
+                            ),
+                        )
+                    })?;
```
Comment (bobozaur) [inline aries_vcx/src/common/proofs/proof_request.rs:59]: Maybe it'd be more idiomatic to create an enum with untagged deserialization trying the `HashMap` first and then `Vec`.

File: aries_vcx/src/common/proofs/proof_request.rs [hunk 3]
```diff
@@ -60,6 +71,33 @@ impl ProofRequestData {
...
+                                "Invalid Requested Predicates: {:?}, err: {:?}",
+                                requested_predicates, err
+                            ),
+                        )
+                    })?;
+                self = self.set_requested_predicates_as_vec(requested_predicates)?;
+            }
+        }
+        Ok(self)
+    }
```
Comment (bobozaur) [inline aries_vcx/src/common/proofs/proof_request.rs:99]: Pretty much same things as for the attributes.

**Review by bobozaur — CHANGES_REQUESTED** [review]
Left some comments about readability mainly.

File: aries_vcx/src/common/proofs/proof_request.rs (hunk 2 above)
Comment (Patrik-Stas) [reply aries_vcx/src/common/proofs/proof_request.rs:59]: The vector format is legacy and deprecated, will go away. I don't think it's worthy improving readability in this case

File: aries_vcx/src/common/proofs/proof_request.rs (hunk 3 above)
Comment (Patrik-Stas) [reply aries_vcx/src/common/proofs/proof_request.rs:99]: dtto

**Review by bobozaur — APPROVED** [review]
Had a short discussion with @Patrik-Stas who pointed out this is kind of legacy code and will most likely get refactored soon.

Since previous comments were mainly about readability, I have no problem with merging this as-is now.

**Pull request merged**